src/parser/on_song_grammar.pegjs

```diff
@@ -0,0 +1,191 @@
+    items:SectionBody*
+    "{" _ ("end_of_" / "eo") DirectiveSectionTypes _ "}" EOL // FIXME: must match start tag
+    {
+      return { type: 'section', name: name || type, items }
+    }
+
+DirectiveSectionTypes
+  = "b" "ridge"? { return "bridge" }
+  / "c" "horus"? { return "chorus" }
+  / "p" "art"?   { return "part" }
+  / "v" "erse"?  { return "verse" }
+
+SectionName
+  = name:MetaName _ ":" EOL {
+    return name.trim()
+  }
+
+SectionBody
+  = !SectionName @(Tab / Stanza) __
+
+Stanza
+  = lines:Line+ {
+    return { type: 'stanza', lines }
+  }
+
+Line
+  = parts:(@ChordsOverLyrics / @(MusicalInstruction / ChordLyricsPair)+ EOL) {
+    return {type: 'line', parts }
+  }
+
+// The other way to express chords in lyrics is to place the chords on a line above the lyrics and
+// use space characters to align the chords with lyrics. This is supported since most music found
+// in other formats use this technique. Here is an example of chords over lyrics:
+//
+// Verse 1:
+//         D           G        D
+// Amazing Grace, how sweet the sound,
+//                          A7
+// That saved a wretch like me.
+//   D                  G      D
+// I once was lost, but now am found,
+//                A7    D
+// Was blind, but now I see.
+ChordsOverLyrics
+  = chords:(_ @(chord:Chord { return { chord, column: location().start.column } }))+ EOL
+    lyrics:(@Lyrics EOL)? {
+      // FIXME:
+      // Is there a better way to do this in PEG?
+      // Or, is there a more idiomatic way to merge chords and lyrics into pairs?
+
+      // First chord does not start at beginning of line, add an empty chord
+      if(chords[0]?.column > 1) chords.unshift({chord: '', column: 1})
+
+      // Ensure lyrics are a string
+      if(!lyrics) lyrics = ''
+
+      const items = [];
+
+      for (let index = 0; index < chords.length; index++) {
+        const { chord, column } = chords[index];
+        const startColumn = column - 1;
+        const endColumn = chords[index + 1]?.column - 1 || lyrics.length;
+        const l = lyrics.padEnd(endColumn, ' ').slice(startColumn, endColumn)
+
+        items.push({type: 'ChordLyricsPair', chords: chord, lyrics: l})
+      }
+
+      return items;
+    }
+
+ChordLyricsPair
+  = chords:BracketedChord lyrics:Lyrics {
+      return { type: "ChordLyricsPair", chords: chords || '', lyrics }
+    }
+  / chords:BracketedChord {
+      return { type: "ChordLyricsPair", chords, lyrics: "" }
+    }
+  / lyrics:Lyrics {
+    return { type: "ChordLyricsPair", lyrics, chords: "" }
+  }
+
+BracketedChord "chord"
+  = !Escape "[" @Chord "]"
+  / !Escape "[" chord:$[^\]]+ "]" {
+      warn(`Unknown chord: ${chord}`, location())
+    }
+
+
+// OnSong recognizes chords using the following set of rules:
+// https://www.onsongapp.com/docs/features/formats/onsong/chords/#chords-over-lyrics
+Chord
+  = $(ChordLetter SharpOrFlat? ChordModifier? ChordNumericPosition? ("/" ChordLetter SharpOrFlat?)?)
```

Comment:
This definition from OnSong is insufficient.  I ran this parser against a collection of ~200 sample songs found in the wild and get ~1800 errors. Here's a sampling of chords that fail to parse with this definition: `Dm7(b5)`,  `E7b13`, `B7b5`, ` C9(11)`, `Dm7(9)`.

The [chords over lyrics](https://www.onsongapp.com/docs/features/formats/onsong/chords/#chords-over-lyrics) syntax for OnSong requires being able to identify valid chords. While I think it makes sense to define the parsing in PEG for chord syntax, I don't think it makes sense to have multiple implementations (here and in chord.js). Ideally there would be a `chord.grammar` that gets imported into both this and `chord.js` implementation for parsing, but peg.js/peggy [doesn't support shared grammars yet](https://github.com/peggyjs/peggy/discussions/239).

For now I may replace this with a call to `Chord.parse` to determine if it's a valid chord. It will probably have some performance implications, but I think it's worth it for correctness for now.

@martijnversluis I'd love your thoughts on how to handle this.  I see you just did some work in #460 related to suffixes.

Reply:
> For now I may replace this with a call to `Chord.parse` to determine if it's a valid chord. It will probably have some performance implications, but I think it's worth it for correctness for now.

I just tested this, and it looks like `Chord.parse` doesn't do strict validation on the suffixes. Any word that starts with a `[A-G]` (like `Chord` or `End`) will parse without errors.

Reply:
Yeah I guess the chord sheet parser is currently way more struct than the chord parser is. We could try to convert the [regexes used for parsing chords](https://github.com/martijnversluis/ChordSheetJS/blob/master/src/chord.js#L20) to grammar? I'll have a thought about it, addressing this should probably be a separate PR.